addons/pkg-npm/jaxrs/src/main/java/org/commonjava/indy/pkg/npm/jaxrs/NPMContentAccessHandler.java

```diff
@@ -0,0 +1,498 @@
+                versionContent = entry.getValue().toString();
+            }
+
+            if ( anode.fields().hasNext() )
+            {
+                String tarball = anode.fields().next().getKey();
+                if ( tarball == null )
+                {
+                    return;
+                }
+                tarballPath = Paths.get( idnode.asText(), "-", tarball ).toString();
+                tarballContent = anode.findPath( "data" ).asText();
+            }
+
+            if ( versionPath == null || tarballPath == null )
+            {
+                return;
+            }
+            versionTarget = transfers.getCacheReference( new ConcreteResource( resource.getLocation(), versionPath ) );
+            logger.info( "STORE {}", versionTarget.getResource() );
+
+            tarballTarget = transfers.getCacheReference( new ConcreteResource( resource.getLocation(), tarballPath ) );
+            logger.info( "STORE {}", tarballTarget.getResource() );
+        }
+        catch ( final IOException e )
+        {
+            logger.error( String.format( "[NPM] Json node parse failed for resource: %s. Reason: %s", resource,
+                                         e.getMessage() ), e );
+            response = formatResponse( e, builderModifier );
+
+            return;
+        }
+
+        if ( versionTarget == null || tarballTarget == null )
+        {
+            return;
+        }
+
+        try (OutputStream versionOutputStream = versionTarget.openOutputStream( TransferOperation.UPLOAD, true,
+                                                                                eventMetadata );
+             OutputStream tarballOutputStream = tarballTarget.openOutputStream( TransferOperation.UPLOAD, true,
+                                                                                eventMetadata ))
+        {
+            versionOutputStream.write( versionContent.getBytes() );
+            tarballOutputStream.write( Base64.decodeBase64( tarballContent ) );
+            generatedTransfers = new GeneratedTransfers( transfer, versionTarget, tarballTarget );
+        }
+        catch ( final IOException e )
+        {
+            logger.error( String.format( "[NPM] Failed to store the generated targets: s% and s%. Reason: s%",
+                                         versionTarget.getResource(), tarballTarget.getResource(), e.getMessage() ),
+                          e );
+            response = formatResponse( e, builderModifier );
+        }
+    }
+
+    public void generateHttpMetadataHeaders( final Transfer transfer, final HttpServletRequest request,
+                                             final Response response )
+    {
+        if ( transfer == null || !transfer.exists() || request == null || response == null )
+        {
+            return;
+        }
+
+        Response responseWithLastModified =
+                response.fromResponse( response ).lastModified( new Date( transfer.lastModified() ) ).build();
```

Comment:
Please call static method from class level but not instance level.

Reply:
@gli, sorry I’m not very clear about this suggestion, I re-wrap the response with lastModified filed because npm requires a strict filed of this, or will result some header check error, do you mean we have other ways simply to implement this instead of ‘response.fromResponse’?

Reply:
as fromResponse is a static method, so directly use Response.fromResponse (class level) is preferred.